### PR TITLE
Generalized path

### PR DESCRIPTION
**Pull request by imkiva**
This PR added _generalized path type_ to aya and refactored the internal representation of partial elements.

### CODE REVIEW

**Comment by imkiva** [comment]
Path application!!!!!111
```
def cong
  {A B : Type}
  (f : A -> B)
  (a b : A)
  (p : a = b)
  : --------------
  f a = f b 
  => \i => f (p i)

def funExt
  {A B : Type}
  (f g : A -> B)
  (p : Pi (a : A) -> f a = g a)
  : ---------------------------------
  f = g 
  => \i a => p a i

```

File: base/src/main/java/org/aya/core/visitor/Expander.java [hunk 1]
```diff
@@ -115,6 +116,19 @@ record Normalizer(@NotNull TyckState state) implements Expander {
         case ElimTerm.App app -> applyThoroughly(CallTerm::make, app);
         case ElimTerm.Proj proj -> ElimTerm.proj(proj);
         case IntroTerm.PartEl el -> partial(el);
+        case ElimTerm.PathApp app -> {
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/visitor/Expander.java:119]: This should be in `Expander`

File: base/src/main/java/org/aya/tyck/ExprTycker.java [hunk 2]
```diff
@@ -388,28 +395,65 @@ else if (iter.hasNext()) {
...
+            var body = dt.substBody(resultParam.toTerm());
+            yield localCtx.with(resultParam, () -> {
+              var rec = inherit(lam.body(), body);
+              return new Result(new IntroTerm.Lambda(resultParam, rec.wellTyped), dt);
+            });
+          }
+          // Path lambda!
+          case FormTerm.Path path -> {
+            var cubeParams = path.cube().params();
+            var plam = Expr.unPathLam(lam, cubeParams.size());
```
Comment (ice1000) [inline base/src/main/java/org/aya/tyck/ExprTycker.java:423]: Beware that there is a minor thing we can improve from guest: if the term is not a lam but can be eta-expand to a lam, then this code will fail

File: base/src/main/java/org/aya/tyck/LittleTyper.java [hunk 3]
```diff
@@ -80,6 +84,18 @@ public record LittleTyper(@NotNull TyckState state, @NotNull LocalCtx localCtx)
       case FormTerm.PartTy ty -> FormTerm.Univ.ZERO;
       case IntroTerm.HappyPartEl el -> new FormTerm.PartTy(el.rhsType(), el.restr());
       case IntroTerm.SadPartEl el -> new FormTerm.PartTy(term(el), el.restr());
+      case FormTerm.Path path -> FormTerm.Univ.ZERO;
+      case IntroTerm.PathLam lam -> new FormTerm.Path(new Cube<>(
+        lam.params().map(Term.Param::ref),
+        term(lam.body()),
+        ImmutableSeq.empty() // TODO: clauses???
```
Comment (ice1000) [inline base/src/main/java/org/aya/tyck/LittleTyper.java:91]: I think `FormTerm.Path` should essentially contain a `PartEl` not a list of clauses. This means we can have trivial `Path`s, which is desirable over here.

File: base/src/main/java/org/aya/core/visitor/Expander.java (hunk 1 above)
Comment (imkiva) [reply base/src/main/java/org/aya/core/visitor/Expander.java]: Yes!!!

Comment (imkiva) [reply base/src/main/java/org/aya/core/visitor/Expander.java]: image

File: base/src/main/java/org/aya/tyck/LittleTyper.java (hunk 3 above)
Comment (imkiva) [reply base/src/main/java/org/aya/tyck/LittleTyper.java]: Now the path type `Cube` holds a generalized partial element `Partial` which can be either happy or sad.

File: base/src/main/java/org/aya/tyck/ExprTycker.java [hunk 4]
```diff
@@ -388,28 +404,62 @@ else if (iter.hasNext()) {
...
+            if (!plam._1.sizeEquals(cubeParams))
+              yield fail(lam, term, new CubicalProblem.DimensionMismatch(lam, cubeParams.size(), plam._1.size()));
+            // we allow lambda params to be typed explicitly --- check them against I.
+            var params = plam._1.map(p -> {
+              inherit(p.type(), FormTerm.Interval.INSTANCE);
```
Comment (ice1000) [inline base/src/main/java/org/aya/tyck/ExprTycker.java:437]: Are you sure `inherit` instead of `unify`?

Comment (imkiva) [reply base/src/main/java/org/aya/tyck/ExprTycker.java:437]: Oh! nice catch!

File: base/src/main/java/org/aya/tyck/ExprTycker.java [hunk 5]
```diff
@@ -434,8 +434,9 @@ yield switch (term.normalize(state, NormalizeMode.WHNF)) {
               yield fail(lam, term, new CubicalProblem.DimensionMismatch(lam, cubeParams.size(), plam._1.size()));
             // we allow lambda params to be typed explicitly --- check them against I.
             var params = plam._1.map(p -> {
-              inherit(p.type(), FormTerm.Interval.INSTANCE);
-              return new Term.Param(p, FormTerm.Interval.INSTANCE);
+              var i = synthesize(p.type());
+              unifyTyReported(FormTerm.Interval.INSTANCE, i.wellTyped, p.type());
```
Comment (ice1000) [inline base/src/main/java/org/aya/tyck/ExprTycker.java:438]: Isn't this the same thing?

Comment (imkiva) [reply base/src/main/java/org/aya/tyck/ExprTycker.java]: No. `inherit(p.type(), I)` says `p.type() : I` while we expect `p.type() = I`

**Comment by ice1000** [comment]
Great!!

**Comment by imkiva** [comment]
bors d=ice1000

**Comment by ice1000** [comment]
I'm curious on why this has reduced coverage. I'll check it out later today.

**Comment by ice1000** [comment]
Currently it's 
image

**Comment by ice1000** [comment]
Ok, -0.27% is a small win
image

File: base/src/main/java/org/aya/core/serde/Serializer.java [hunk 6]
```diff
@@ -111,11 +111,12 @@ public record Serializer(@NotNull Serializer.State state) {
...
-      case FormTerm.PartTy ty -> throw new UnsupportedOperationException("TODO");
-      case FormTerm.Path path -> throw new UnsupportedOperationException("TODO");
-      case IntroTerm.PathLam path -> throw new UnsupportedOperationException("TODO");
-      case ElimTerm.PathApp app -> throw new UnsupportedOperationException("TODO");
+      // TODO: implement these
+      case IntroTerm.PartEl el -> new SerTerm.Univ(114514);
+      case FormTerm.PartTy ty -> new SerTerm.Univ(114514);
+      case FormTerm.Path path -> new SerTerm.Univ(114514);
+      case IntroTerm.PathLam path -> new SerTerm.Univ(114514);
+      case ElimTerm.PathApp app -> new SerTerm.Univ(114514);
```
Comment (imkiva) [inline base/src/main/java/org/aya/core/serde/Serializer.java:119]: LOL

**Pull request merged**